File: Aydinyan/stack.js
var node = function(val, pr = null) {
	this.prev = pr;
	this.value = val;
}

var stack = function() {
	this.length = 0;
	this.last = null;
}

stack.prototype.push = function(data) {
	this.last = new node(data, this.last);
	this.length++;
}

stack.prototype.pop = function() {
    if (this.length > 0) {
        let n = this.last;    
        this.last = n.prev;
        n.prev = null;
        this.length--;
        return n.value;
    }
    else {
        return ;
    }
}

stack.prototype.isEmpty = function() {
	return this.length === 0;
}

stack.prototype.print = function() {
    if (this.length > 0) {
	    let n = this.last;
	    while (n !== null) {
            console.log(n.value);
	    	n = n.prev;
        }
    }
}

stack.prototype.clear = function() {
    this.last = null;
    this.length = 0;
}

stack.prototype.top = function() {
    return this.last.value;
}

stack.prototype.size = function() {
    return this.length;
}

let s = new stack();
s.push(12);
s.push(1);
s.push(2);
s.print();
console.log(s.top());
console.log('\n')
console.log(s.pop());
console.log('\n')
s.print();
console.log(s.isEmpty());
s.clear();
s.print();
console.log(s.isEmpty());

File: Hrach/stack.dart
class node {
  node(value) {
    _prev = null;
    _value = value;
  }

  node _prev;
  var _value;
}

class stack {
  stack() {
    _vertex = null;
    _length = 0;
  }

  void push(var value) {
    node tmp = node(value);
    if (isEmpty()) {
      _vertex = tmp;
    } else {
      tmp._prev = _vertex;
      _vertex = tmp;
    }
    _length++;
  }

  dynamic pop() {
    if (isEmpty()) {
      return null;
    }
    var val = _vertex._value;
    _vertex = _vertex._prev;
    _length--;
    return val;
  }

  dynamic top() {
    return isEmpty() ? null : _vertex._value;
  }

  void erase() {
    _vertex = null;
    _length = 0;
  }

  bool isEmpty() {
    return _length == 0;
  }

  void printAll() {
    node tmp = _vertex;
    while (tmp != null) {
      print(tmp._value);
      tmp = tmp._prev;
    }
  }

  get length => _length;
  node _vertex;
  int _length;
}
